File: JvmClient/src/commonMain/kotlin/spaceEngineers/controller/BlocksAdmin.kt
package spaceEngineers.controller

import spaceEngineers.model.BlockId
import spaceEngineers.model.BlockLocation
import spaceEngineers.model.CubeGrid
import spaceEngineers.model.DefinitionId
import spaceEngineers.model.Vec3F
import spaceEngineers.model.Vec3I

interface BlocksAdmin {

    fun placeAt(
        blockDefinitionId: DefinitionId,
        position: Vec3F,
        orientationForward: Vec3F = Vec3F.FORWARD,
        orientationUp: Vec3F = Vec3F.UP,
        color: Vec3F? = null,
    ): CubeGrid

    fun placeInGrid(
        blockDefinitionId: DefinitionId,
        gridId: String,
        minPosition: Vec3I,
        orientationForward: Vec3I = Vec3I.FORWARD,
        orientationUp: Vec3I = Vec3I.UP,
        color: Vec3F? = null,
    ): String

    fun batchPlaceInGrid(
        gridId: String,
        color: Vec3F? = null,
        blockPlacementConfigs: List<BlockLocation>
    ): List<BlockId>

    fun remove(blockId: BlockId)
    fun setIntegrity(blockId: BlockId, integrity: Float)
    fun setCustomName(blockId: BlockId, customName: String)
    fun createOrUpdateGroup(name: String, gridId: String, blockIds: List<BlockId>)
    fun mapButtonToBlock(buttonBlockId: BlockId, buttonIndex: Int, action: String, targetId: BlockId)
    fun mapButtonToGroup(buttonBlockId: BlockId, buttonIndex: Int, action: String, groupName: String)
    fun setPhysicalItemInInventory(blockId: BlockId, definitionId: DefinitionId, amount: Float)
    fun requestConversionToShip(gridId: String)
    fun requestConversionToStation(gridId: String)
    fun setDestructibleBlocks(gridId: String, destructibleBlocks: Boolean)

    val warhead: WarheadAdmin
    val medicalRoom: MedicalRoomAdmin
    val functionalBlock: FunctionalBlockAdmin
    val terminalBlock: TerminalBlockAdmin
    val sensorBlock: SensorBlockAdmin
    val pistonBase: PistonBaseAdmin
    val thrust: ThrustAdmin
    val timerBlock: TimerBlockAdmin
    val buttonPanel: ButtonPanelAdmin
    val door: DoorBaseAdmin
}

interface DoorBaseAdmin {
    fun setAnyoneCanUse(blockId: String, anyoneCanUse: Boolean)
    fun setOpen(blockId: String, open: Boolean)
}

interface ButtonPanelAdmin {
    fun setAnyoneCanUse(blockId: String, anyoneCanUse: Boolean)
}

interface TimerBlockAdmin {
    fun setTriggerDelay(blockId: BlockId, triggerDelay: Float)
    fun start(blockId: BlockId)
    fun stop(blockId: BlockId)
    fun triggerNow(blockId: BlockId)
}

interface ThrustAdmin {
    fun setThrustOverride(blockId: BlockId, thrustOverride: Float)
}

interface PistonBaseAdmin {
    fun setVelocity(blockId: String, velocity: Float)
    fun recreateTop(blockId: String)
}

interface SensorBlockAdmin {
    fun setFieldMin(blockId: String, fieldMin: Vec3F)
    fun setFieldMax(blockId: String, fieldMax: Vec3F)
}

interface MedicalRoomAdmin {
    fun setRespawnAllowed(blockId: String, respawnAllowed: Boolean)
    fun setHealingAllowed(blockId: String, healingAllowed: Boolean)
    fun setRefuelAllowed(blockId: String, refuelAllowed: Boolean)
    fun setSpawnWithoutOxygenEnabled(blockId: String, spawnWithoutOxygenEnabled: Boolean)
}

interface WarheadAdmin {
    fun explode(blockId: BlockId)
    fun detonate(blockId: BlockId)
    fun startCountDown(blockId: BlockId): Boolean
    fun stopCountdown(): Boolean
    fun setArmed(blockId: BlockId, armed: Boolean)
}

interface FunctionalBlockAdmin {
    fun setEnabled(blockId: BlockId, enabled: Boolean)
}

interface TerminalBlockAdmin {
    fun setCustomName(blockId: BlockId, customName: String)
    fun setCustomData(blockId: BlockId, customData: String)
    fun setShowInInventory(blockId: String, showInInventory: Boolean)
    fun setShowInTerminal(blockId: String, showInTerminal: Boolean)
    fun setShowOnHUD(blockId: String, showOnHUD: Boolean)
}

File: Source/Ivxr.SpaceEngineers/BlocksAdmin.cs
using System.Collections.Generic;
using Iv4xr.SpaceEngineers.WorldModel;

namespace Iv4xr.SpaceEngineers
{
    public interface IBlocksAdmin
    {
        void Remove(string blockId);

        void SetIntegrity(string blockId, float integrity);
        void SetCustomName(string blockId, string customName);
        void CreateOrUpdateGroup(string name, string gridId, List<string> blockIds);
        void MapButtonToBlock(string buttonBlockId, int buttonIndex, string action, string targetId);
        void MapButtonToGroup(string buttonBlockId, int buttonIndex, string action, string groupName);

        CubeGrid PlaceAt(DefinitionId blockDefinitionId, PlainVec3D position, PlainVec3D orientationForward,
            PlainVec3D orientationUp, PlainVec3F? color);

        string PlaceInGrid(DefinitionId blockDefinitionId, string gridId, PlainVec3I minPosition,
            PlainVec3I orientationForward, PlainVec3I orientationUp, PlainVec3F? color);

        List<string> BatchPlaceInGrid(string gridId, PlainVec3F? color, List<BlockLocation> blockPlacementConfigs);
        void SetPhysicalItemInInventory(string blockId, DefinitionId definitionId, float amount);
        void RequestConversionToShip(string gridId);
        void RequestConversionToStation(string gridId);
        void SetDestructibleBlocks(string gridId, bool destructibleBlocks);

        IWarheadAdmin Warhead { get; }
        IFunctionalBlockAdmin FunctionalBlock { get; }
        ITerminalBlockAdmin TerminalBlock { get; }
        IMedicalRoomAdmin MedicalRoom { get; }
        ISensorBlockAdmin SensorBlock { get; }
        IGravityGeneratorAdmin GravityGenerator { get; }
        IGravityGeneratorSphereAdmin GravityGeneratorSphere { get; }
        ITimerBlockAdmin TimerBlock { get; }
        IPistonBaseAdmin PistonBase { get; }
        IThrustAdmin Thrust { get; }
        IButtonPanelAdmin ButtonPanel { get; }
        IDoorBaseAdmin Door { get; }
        ITextPanelAdmin TextPanel { get; }

    }

    public interface IGravityGeneratorBaseAdmin
    {
        void SetGravityAcceleration(string blockId, float gravityAcceleration);
    }

    public interface IGravityGeneratorAdmin : IGravityGeneratorBaseAdmin
    {
        void SetFieldSize(string blockId, PlainVec3D fieldSize);
    }

    public interface IGravityGeneratorSphereAdmin : IGravityGeneratorBaseAdmin
    {
        void SetRadius(string blockId, float radius);
    }

    public interface ITimerBlockAdmin
    {
        void SetTriggerDelay(string blockId, float triggerDelay);
        void Start(string blockId);
        void Stop(string blockId);
        void TriggerNow(string blockId);
    }

    public interface IWarheadAdmin
    {
        void Explode(string blockId);

        void Detonate(string blockId);

        bool StartCountdown(string blockId);

        bool StopCountdown(string blockId);

        void SetArmed(string blockId, bool armed);
        void SetCountdownMs(string blockId, int countdown);
    }

    public interface IFunctionalBlockAdmin
    {
        void SetEnabled(string blockId, bool enabled);
    }

    public interface ITerminalBlockAdmin
    {
        void SetCustomName(string blockId, string customName);
        void SetCustomData(string blockId, string customData);
        void SetShowInInventory(string blockId, bool showInInventory);
        void SetShowInTerminal(string blockId, bool showInTerminal);
        void SetShowOnHUD(string blockId, bool showOnHUD);
    }

    public interface IMedicalRoomAdmin
    {
        void SetRespawnAllowed(string blockId, bool respawnAllowed);
        void SetHealingAllowed(string blockId, bool healingAllowed);
        void SetRefuelAllowed(string blockId, bool refuelAllowed);
        void SetSpawnWithoutOxygenEnabled(string blockId, bool spawnWithoutOxygenEnabled);
    }

    public interface ISensorBlockAdmin
    {
        void SetFieldMin(string blockId, PlainVec3D fieldMin);
        void SetFieldMax(string blockId, PlainVec3D fieldMax);
    }

    public interface IDoorBaseAdmin
    {
        void SetAnyoneCanUse(string blockId, bool anyoneCanUse);
        void SetOpen(string blockId, bool open);
    }
    
    public interface IButtonPanelAdmin
    {
        void SetAnyoneCanUse(string blockId, bool anyoneCanUse);
    }

    public interface IPistonBaseAdmin
    {
        void SetVelocity(string blockId, float velocity);
        void RecreateTop(string blockId);
    }

    public interface ITextPanelAdmin
    {
        void SetPublicTitle(string blockId, string publicTitle);
        void SetPrivateTitle(string blockId, string privateTitle);
        void SetPrivateDescription(string blockId, string privateDescription);
        void SetContentType(string blockId, int contentType);
        void SetTextPadding(string blockId, float padding);
        void SetText(string blockId, string text);
        void SetAlignment(string blockId, int alignment);
        void SetFontSize(string blockId, float fontSize);
    }

    public interface IThrustAdmin
    {
        void SetThrustOverride(string blockId, float thrustOverride);
    }

    public interface ICharacterAdmin
    {
        CharacterObservation Teleport(PlainVec3D position, PlainVec3D? orientationForward = null,
            PlainVec3D? orientationUp = null);

        void Use(string blockId, int functionIndex, int action);

        CharacterObservation Create(string name, PlainVec3D position, PlainVec3D orientationForward,
            PlainVec3D orientationUp);

        void Switch(string id);
        void Remove(string id);
        void ShowTerminal(string blockId);
        void Die();
        string MainCharacterId();
        string LocalCharacterId();
        void UpdateEnergy(float energy);
        void UpdateOxygen(float oxygen);
        void UpdateHydrogen(float hydrogen);
    }
}
